src/Native/LibTorchSharp/THSJIT.cpp

```diff
@@ -145,45 +155,174 @@ JITMethod THSJIT_Module_get_method(const JITModule module, const char* name)
     return new std::shared_ptr<torch::jit::Method>(copy);
 }
 
-void THSJIT_Module_forward(const JITModule module, const Tensor* tensorPtrs, const int length, Tensor* (*allocator)(size_t length), int8_t* typeCode)
+void ReturnHelper(c10::IValue result, TensorOrScalar* (*allocator)(size_t length), int8_t* typeCode)
 {
-    *typeCode = 0;
-
-    CATCH(
-        auto result = (*module)->forward(toTensors<c10::IValue>((torch::Tensor**)tensorPtrs, length));
-
     // TypeCode:
-    //
-    // 0 -- Not supported
-    // 1 -- Single tensor
-    // 2 -- Tuple of tensors
-    // 3 -- List of tensors
+//
+// 0 -- Not supported
+// 1 -- Single tensor
+// 2 -- Tuple of tensors
+// 3 -- List of tensors
+// 4 -- Single scalar
+// 5 -- Scalar tuple
+// 6 -- List of scalars
+// 7 -- List of scalars and tensors
+
+    if (result.isScalar())
+    {
+        TensorOrScalar* output = allocator(1);
+        output[0] = { 0, (ptrdiff_t)new torch::Scalar(result.toScalar()) };
+        *typeCode = 4;
+        return;
+    }
 
     if (result.isTensor()) {
-        Tensor* output = allocator(1);
-        output[0] = ResultTensor(result.toTensor());
+        TensorOrScalar* output = allocator(1);
+        output[0] = { 0, (ptrdiff_t)ResultTensor(result.toTensor()) };
         *typeCode = 1;
         return;
     }
+
     if (result.isTensorList()) {
         auto list = result.toTensorList();
         *typeCode = 3;
-        Tensor* output = allocator(list.size());
+        TensorOrScalar* output = allocator(list.size());
         for (size_t i = 0; i < list.size(); i++)
-            output[i] = ResultTensor(list[i]);
+            output[i] = { 0, (ptrdiff_t)ResultTensor(list[i]) };
         return;
     }
+
+    if (result.isList())
+    {
+        int foundTensor = 0;
+        int foundScalar = 0;
+
+        auto list = result.toList();
+        TensorOrScalar* output = allocator(list.size());
+
+        for (int i = 0; i < list.size(); ++i)
+        {
+            output[i].Handle = -1;
+            c10::IValue value = list[i];
+
+            if (value.isTensor())
```

Comment:
>value [](http://example.com/codeflow?start=16&length=5)

is it possible the list can have tuple items? I mean is it possible we can run into the case that have value.isTuple() is true?

Reply:
Yes, and we'll reject those for now.